designdocs/state_storage.md

```diff
@@ -0,0 +1,101 @@
+# On Chain State Storage
+
+## Context and Scope
+The proposed feature consists of changes that is going to be one (or more) pallet(s) in runtime of a
+Substrate based blockchain, and it will be used in all environments including production.
+
+## Problem Statement
+Every party using any medium of communication will need to deal with the storage of some kind of
+stateful data. Some of these stateful data might be consensus critical and in a decentralized world
+one option might be to store this type of data on a blockchain as long as it is required.
+The scope for how long some of these stateful data should exist is dependent on the
+type of data. For example, some of them might exist as long as an active channel exists between two
+communicating parties and for some others it might be directly dependent on the existence of a
+particular party.
+
+### DSNP Usecase
+**Frequency** is the first implementation of [DSNP](https://dsnp.org) and most of the state
+transitions in **DSNP** can be modeled via `Announcements`. For some of these state transitions we
+only care about the latest state, and currently the only way to achieve this via announcements is to
+have some kind of third party indexer to track the latest state for these types.
+If there was a way to be able to mutate the state of a storage, then we could update it to its
+latest version and potentially remove the need for an indexer.
+
+## Goals
+1. Allowing storage of stateful data with flexible schemas on chain.
+2. Data stored for any `MSA` should not have any effects like changing read/write access times or
+storage costs for any data stored for another `MSA`.
+3. Allowing high write throughput.
+4. Allowing high read throughput.
+5. Avoiding data races between consecutive updates of the same item.
+
+## Non-goals
+- Defining an economic incentive to restrict infinite growth of on-chain database.
+
+## Glossary
+* `DSNP`: Decentralized Social Networking Protocol
+* `MSA`: Message Source Account. A registered identifier with the MSA pallet.
+
+## Proposal
+Creating a double map of (`SchemaId` and `PageNumber`) as keys and `vec<u8>` as stored values.
+
+Using mentioned double-map will allow storing data using a variety of schemas using pagination.To
+be able to achieve goal number 2 we will need to create a child-tree for each `MSA` and store this
+double map under these child trees for each `MSA`.
+
+### Modes of Storage
+- **Batch**: In this mode we are storing each entity of that schemaId as an item inside an array.
+This is recommended for items that are relatively small and there are more than a few of them.
+- **Single**: In this mode we are storing each entity individually, and it is recommended for items
```

Comment:
Double checking that I understand (and clear to others). Single refers to a single page. We are not trying to ever have one schema that is combining pages. Is that a correct statement?

Reply:
That is correct, each item is in it's own page.